R/Chrome.R

```diff
@@ -2,6 +2,141 @@
+#' @param local Logical scalar, indicating whether the local version of the
+#'     protocol (embedded in `crrri`) must be used or the protocol must be
+#'     fetched _remotely_.
+#' @param extra_args Character vector, extra command line arguments passed to
+#'     Chromium/Chrome.
+#' @param headless Logical scalar, indicating whether Chromium/Chrome is launched
+#'     in headless mode.
+#' @param retry_delay Number, delay in seconds between two successive tries to
+#'     connect to headless Chromium/Chrome.
+#' @param max_attempts Logical scalar, number of tries to connect to headless
+#'     Chromium/Chrome.
+#'
+#' @return An invisible list with the values of the fulfilled promises for each
+#'     async function.d If there is only async function, the return value is the value of the
+#'     fulfilled promise.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' async_save_as_pdf <- function(url) {
+#'   function(client) {
+#'     Page <- client$Page
+#'
+#'     Page$enable() %...>% {
+#'       Page$navigate(url = url)
+#'       Page$loadEventFired()
+#'     } %...>% {
+#'       Page$printToPDF()
+#'     } %...>% {
+#'       .$data %>%
+#'         jsonlite::base64_dec() %>%
+#'         writeBin(paste0(httr::parse_url(url)$hostname, ".pdf"))
+#'     }
+#'   }
+#' }
+#'
+#' save_as_pdf <- function(...) {
+#'   list(...) %>%
+#'     purrr::map(async_save_as_pdf) %>%
+#'     chrome_execute(.list = .)
+#' }
+#'
+#' save_as_pdf("https://www.r-project.org/", "https://rstudio.com/")
+#' }
+chrome_execute <- function(
+  ..., .list = NULL, timeouts = 30, cleaning_timeout = 30, async = FALSE,
+  bin = Sys.getenv("HEADLESS_CHROME"), debug_port = 9222L, local = FALSE,
+  extra_args = NULL, headless = TRUE, retry_delay = 0.2, max_attempts = 15L
+) {
+  if(is.null(.list)) {
+    funs <- list(...)
+  } else {
+    stopifnot(is.list(.list))
+    funs <- .list
+  }
+
+  # check arguments
+  purrr::walk(funs, check_is_single_param_fun)
+  stopifnot(is.numeric(timeouts))
+  assert_that(is.number(cleaning_timeout))
+
+  # initialize objects
+  timeouts <- rep_len(timeouts, length(funs))
+  total_timeout <- sum(timeouts) + cleaning_timeout
+  results <- vector("list", length(funs))
+  curr_env <- rlang::current_env()
+  execute_fun <- function(client, index, env) {
+    fun <- purrr::pluck(funs, index)
+    delay <- purrr::pluck(timeouts, index)
+    result_saved <- promises::then(fun(client), onFulfilled = function(value) {
+      if(!is.null(value)) {
+        results <- rlang::env_get(env, "results")
+        rlang::env_bind(env, results = purrr::assign_in(results, index, value))
+      }
+      client
+    })
+    timeout(result_saved, delay = delay)
+  }
+
+  # launch Chrome
+  chrome <- Chrome$new(
+    bin = bin, debug_port = debug_port, local = local, extra_args = extra_args,
+    headless = headless, retry_delay = retry_delay, max_attempts = max_attempts
+  )
+  client <- chrome$connect()
+  all_funs_executed <- promises::promise_reduce(
```

Comment:
`promises::promise_reduce` works here because your `execute_fun` return `client` right ? 

As the `funs` are independant to one another, I would have expecter a `promise_map` here. 
`promise_map` execute sequentially to and return a list with the values. Could it be useful in some case? 

The trick is we don't know the argument name of the `funs` unless we force user to name it `client`. 
I would have imagine something like
```r
promise_map(funs, execute_fun, client = client)
```

Did you try something like that ?

Reply:
When I began working on that, I forgot that `promises::promise_map()` was sequential and I wanted to avoid concurrency on the same target. I thought that the only way was to use `promise_reduce`.
Once finished, I re-read the doc and realized that `promises::promise_map()` should do the job (if the requirement that `funs` return a promise is respected by the user, otherwise there would be some concurrency). 
I think this would be better and much more compact.

However, I am not sure to understand the problem of not naming the `funs` argument and why we should force user to name it `client`. This is basically a `do.call()`, isn't it?

Reply:
> However, I am not sure to understand the problem of not naming the funs argument and why we should force user to name it client. It is basically a call, isn't it?

I may have thought to quickly. You need to provide named argument in ˋpromise_map` for argument you are not iterating on but It seems ok here. 
I’ll try tomorrow.

Reply:
Done in 60670ea27397d0a78b0a344f6130fa434b8680b8